{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Readme\n",
    "\n",
    "This script contains codes for data simulation and method evaluation described in the manuscript titled 'Benchmarking algorithms for joint integration of unpaired and paired single-cell RNA-seq and ATAC-seq data'. \n",
    "\n",
    "There are five major scenarios and one real data scenario described in this script, designed to answer three questions related to single-cell multi-omic integration. \n",
    "\n",
    "Scenario 1 tests if the multiome dataset helps the integration of unpaired scRNA-seq and snATAC-seq dataset. \n",
    "\n",
    "Scenario 2 tests how sequencing depth of the multiome dataset influences integration result and compare methods at their performance on cell type annotation, peak-gene recovery, cell type mixing, and batch mixing.  \n",
    "\n",
    "Scenario 3 compares methods when the three data types are collected from different donor or research site, in other words, challenges each method at its ability to combat batch effect. \n",
    "\n",
    "Scenario 4 compares methods when there are cell population(s) missing from one of the single-modality dataset (scRNA-seq or snATAC-seq). Thus, testing each method's ability to integrate datasets in the case of unsahred cell populations. \n",
    "\n",
    "Scenario 5 compares methods when there are cell population(s) missing from the multiome dataset or exclusive to one of the single-modality dataset (scRNA-seq or snATAC-seq). Thus, testing each method's ability to integrate datasets in another type of unsahred cell population scenario. \n",
    "\n",
    "Section 6 applies the methods to a real data scenario where multiome, scRNA-seq, and snATAC-seq datasets from the Human Pancreas Analysis Program are integrated. \n",
    "\n",
    "In each sceanrio, we created multiple challenges using two publiclly available dataset (PBMC and BMMC). Open this script in Google Colab or Jupyter Notebook for the best visualization, as the codes are divided into 6 sections and 17 challenges in total. \n",
    "\n",
    "For each challenge, there are two parts. First one demonstrates how data is simulated and the parameters used. Second part shows codes used to run all methods and the files used to evaluate integration result. We ran all evaluations on our LPC server, through submitting one non-interactive job for completing one challenge using one method. Please adopt these codes depending on the environment you are running these codes in.  \n",
    "\n",
    "This script assumes that a conda environment has been created for each method.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "rXhERnIMrgZx"
   },
   "outputs": [],
   "source": [
    "# Run with a python kernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "JFmJ7_9TrgZz",
    "outputId": "c37cbef8-42b9-4998-9d3f-5575c1189cc2"
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# load libraries\n",
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from anndata import AnnData\n",
    "import anndata as ad\n",
    "from copy import deepcopy\n",
    "from scipy.sparse import csr_matrix\n",
    "from scipy.sparse.linalg import svds\n",
    "import pickle\n",
    "import scipy.io as sio\n",
    "import os\n",
    "\n",
    "# Path to R [modify as needed]\n",
    "os.environ['R_HOME'] = '/home/myylee/anaconda3/envs/r_py/lib/R/'\n",
    "import utils_eval\n",
    "\n",
    "\n",
    "from data_simulation import pair_unpair_split_size,downsample_samples, data_simulation, data_simulation_batch\n",
    "from data_simulation import eval_test_all\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PBMC "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "source_dir = \"dataset/multiome_pbmc_10k/\"\n",
    "\n",
    "adata_rna = ad.read_h5ad(source_dir+\"pbmc_10x_rna_public.h5ad\")\n",
    "adata_atac = ad.read_h5ad(source_dir+\"pbmc_10x_atac_public.h5ad\")\n",
    "\n",
    "in_dir = \"dataset/multiome_pbmc_10k/pbmc_vary_cell_test/\"\n",
    "\n",
    "iter_list = [1000,3000,8000]\n",
    "\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list)\n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "\n",
    "n_scrna_list = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "\n",
    "repeats = 5\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "fragment_path = \"{}/dataset/multiome_pbmc_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "\n",
    "cond_key = \"nmulti\"\n",
    "    \n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
    "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                repeats,fragment_path,cond_key,\n",
    "                to_str, downsample=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_single.py\"\n",
    "\n",
    "ct_ref = \"dataset/multiome_pbmc_10k/pbmc_10x_bc_ct3.csv\"\n",
    "nclust = 7\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
    "gp_truth = \"dataset/multiome_pbmc_10k/pbmc_10x_pmat_sig_links_50kb_unique.csv\"\n",
    "\n",
    "dir_path = \"dataset/multiome_pbmc_10k/pbmc_vary_cell_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [1000,3000,8000]\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\",\"run_scmomat_pbmc.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*3\n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# ===== unpaired =====\n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3_single.R\",\n",
    "                  \"run_rbindsc_single.R\",\n",
    "                  \"run_rfigr_single.R\",\n",
    "                  \"run_rliger_single.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4 + [True]\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\"\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# GLUE \n",
    "# ===== unpaired =====\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38_single.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4 + [True]\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# GLUE \n",
    "# ===== unpaired =====\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add dashed line in peak-gene recovery percentage with the true paired RNA and ATAC profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy over the true paired RNA-seq data\n",
    "import shutil\n",
    "dir_path = \"dataset/multiome_pbmc_10k/pbmc_vary_cell_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [1000,3000,8000]\n",
    "repeats = 5\n",
    "output_folder=\"results_single_same_cell_number\"\n",
    "\n",
    "for i in range(len(iter_list)):\n",
    "    for j in range(1,repeats+1):\n",
    "        in_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j))\n",
    "        out_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j),output_folder)\n",
    "\n",
    "        os.makedirs(os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\"), exist_ok=True)\n",
    "        # save ATAC to the right location\n",
    "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"), \n",
    "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"barcodes.tsv\"))\n",
    "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"peak.tsv\"), \n",
    "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"peak.tsv\"))\n",
    "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"ATAC_counts.mtx\"), \n",
    "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"counts.mtx\"))\n",
    "\n",
    "        # load in ATAC barcodes \n",
    "        atac_bc = pd.read_csv(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"),header=None)\n",
    "\n",
    "        # get RNA profile of the corresponding cells \n",
    "        in_dir = \"dataset/multiome_pbmc_10k/\"\n",
    "\n",
    "        adata_rna = ad.read_h5ad(in_dir+\"pbmc_10x_rna_public.h5ad\")\n",
    "\n",
    "        idx_sel = atac_bc[0].tolist()\n",
    "        # RNA\n",
    "        adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna[idx_sel,:].X.todense())),\n",
    "                               obs=adata_rna.obs.loc[idx_sel,:],\n",
    "                               var=adata_rna.var,dtype=np.float32)\n",
    "        adata_rna_sel.obs['rna.bc'] = adata_rna_sel.obs_names\n",
    "        # save RNA to the right location\n",
    "        utils_eval.write_adata(adata_rna_sel, os.path.join(out_dir_i,\"truth\",\"predicted\",\"RNA\"),\"RNA\",\"gene\",feature_name='feature',transpose=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run eval_missming_modality_prediction \n",
    "\n",
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_missing_mod_eval.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_single.py\"\n",
    "\n",
    "ct_ref = \"dataset/multiome_pbmc_10k/pbmc_10x_bc_ct3.csv\"\n",
    "nclust = 7\n",
    "\n",
    "dir_path = \"dataset/multiome_pbmc_10k/pbmc_vary_cell_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [1000,3000,8000]\n",
    "repeats = 5\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
    "gp_truth = \"dataset/multiome_pbmc_10k/pbmc_10x_pmat_sig_links_50kb_unique.csv\"\n",
    "method_keys = [\"truth\"]\n",
    "\n",
    "\n",
    "#----- All of these can be random strings, are not used in script, but need not to be empty strings ----\n",
    "conda_envs = [\"truth_na\"]\n",
    "# R script for running the method\n",
    "method_scripts = [\"run_truth.csv\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "file_paths = [\"truth/truth_result.csv\"]\n",
    "#------ END ----- \n",
    "\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BMMC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using pmat dataset, generated from bmmc_process R script \n",
    "source_dir = \"dataset/bmmc/\"\n",
    "\n",
    "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
    "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "adata_atac.var['feature'] = adata_atac.var['features'] \n",
    "adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "\n",
    "# aggregate the selected donors \n",
    "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
    "\n",
    "# Paired - RNA\n",
    "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "# Paired - ATAC\n",
    "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "\n",
    "\n",
    "in_dir = \"dataset/bmmc/bmmc_vary_cell_test/\"\n",
    "iter_list = [1000,2000,4000]\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "repeats = 5\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
    "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "\n",
    "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
    "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                repeats,fragment_path,cond_key,\n",
    "                to_str, downsample=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_single.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
    "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/bmmc_vary_cell_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [1000,2000,4000]\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\",\"run_scmomat_bmmc.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*3\n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# ===== unpaired =====\n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3_single.R\",\n",
    "                  \"run_rbindsc_single.R\",\n",
    "                  \"run_rfigr_single.R\",\n",
    "                  \"run_rliger_single.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4 + [True]\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\"\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# GLUE \n",
    "# ===== unpaired =====\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38_single.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4 + [True]\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# GLUE \n",
    "# ===== unpaired =====\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add dashed line in peak-gene recovery percentage with the true paired RNA and ATAC profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy over the true paired RNA-seq data\n",
    "import shutil\n",
    "dir_path = \"dataset/bmmc/bmmc_vary_cell_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [1000,2000,4000]\n",
    "repeats = 5\n",
    "output_folder=\"results_single_same_cell_number\"\n",
    "\n",
    "for i in range(len(iter_list)):\n",
    "    for j in range(1,repeats+1):\n",
    "        in_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j))\n",
    "        out_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j),output_folder)\n",
    "\n",
    "        os.makedirs(os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\"), exist_ok=True)\n",
    "        # save ATAC to the right location\n",
    "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"), \n",
    "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"barcodes.tsv\"))\n",
    "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"peak.tsv\"), \n",
    "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"peak.tsv\"))\n",
    "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"ATAC_counts.mtx\"), \n",
    "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"counts.mtx\"))\n",
    "\n",
    "\n",
    "        # load in ATAC barcodes \n",
    "        atac_bc = pd.read_csv(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"),header=None)\n",
    "        idx_sel = atac_bc[0].tolist()\n",
    "\n",
    "        # get RNA profile of the corresponding cells \n",
    "        in_dir = \"dataset/bmmc/\"\n",
    "\n",
    "        adata_rna = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "        adata_rna.obs['rna.bc'] = adata_rna.obs_names\n",
    "        adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "        # RNA\n",
    "        adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna[idx_sel,:].X.todense())),\n",
    "                               obs=adata_rna.obs.loc[idx_sel,:],\n",
    "                               var=adata_rna.var,dtype=np.float32)\n",
    "        # save RNA to the right location\n",
    "        utils_eval.write_adata(adata_rna_sel, os.path.join(out_dir_i,\"truth\",\"predicted\",\"RNA\"),\"RNA\",\"gene\",feature_name='feature',transpose=True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run eval_missming_modality_prediction \n",
    "\n",
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_missing_mod_eval.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_single.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
    "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/bmmc_vary_cell_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [1000,2000,4000]\n",
    "repeats = 5\n",
    "\n",
    "method_keys = [\"truth\"]\n",
    "\n",
    "#----- All of these can be random strings, are not used in script, but need not to be empty strings ----\n",
    "conda_envs = [\"truth_na\"]\n",
    "# R script for running the method\n",
    "method_scripts = [\"run_truth.csv\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "file_paths = [\"truth/truth_result.csv\"]\n",
    "#------ END ----- \n",
    "\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SHARE-seq "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_dir = \"dataset/mouse_skin/\"\n",
    "\n",
    "adata_rna = ad.read_h5ad(source_dir+\"mouse_skin_shareseq_rna.h5ad\")\n",
    "adata_atac = ad.read_h5ad(source_dir+\"mouse_skin_shareseq_atac.h5ad\")\n",
    "\n",
    "adata_rna.var['feature'] = deepcopy(adata_rna.var_names)\n",
    "adata_atac.var['feature'] = deepcopy(adata_atac.var_names)\n",
    "\n",
    "adata_rna.obs['ct3'] = deepcopy(adata_rna.obs['celltype'])\n",
    "adata_atac.obs['ct3'] = deepcopy(adata_atac.obs['celltype'])\n",
    "\n",
    "adata_rna.obs['ct3'].to_csv(in_dir+\"mouse_skin_shareseq_bc_ct3.csv\")\n",
    "\n",
    "in_dir = \"dataset/mouse_skin/multiome_ncells_pmat/\"\n",
    "\n",
    "iter_list = [5000,10000,15000]\n",
    "\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [8000]*len(iter_list)\n",
    "n_snatac_list = [8000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "repeats = 5\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "fragment_path = \"{}/dataset/mouse_skin/mouse_skin_shareseq_fragments.tsv.gz\".format(os.getcwd())\n",
    "\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "def str_norm(s):\n",
    "    return(str(int(s*100)))\n",
    "    \n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "data_simulation(in_dir,adata_rna,adata_atac,iter_list,\n",
    "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                repeats,fragment_path,cond_key,\n",
    "                to_str, downsample=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_single.py\"\n",
    "\n",
    "ct_ref = \"dataset/mouse_skin/mouse_skin_shareseq_bc_ct3.csv\"\n",
    "nclust = 22\n",
    "\n",
    "dir_path = \"dataset/mouse_skin/multiome_ncells_pmat/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list =  [15000,10000,5000]\n",
    "repeats = 5\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single_mm10.R\"\n",
    "gp_truth = \"dataset/mouse_skin/mouse_skin_shareseq_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
    "\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\",\"run_scmomat_mouse_skin.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*3\n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# ===== unpaired =====\n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3_single.R\",\n",
    "                  \"run_rbindsc_single.R\",\n",
    "                  \"run_rfigr_single.R\",\n",
    "                  \"run_rliger_single.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4 + [True]\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\"\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# GLUE \n",
    "# ===== unpaired =====\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_mm10_single.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4 + [True]\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# GLUE \n",
    "# ===== unpaired =====\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_mm10.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add dashed line in peak-gene recovery percentage with the true paired RNA and ATAC profiles "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import shutil\n",
    "dir_path = \"dataset/mouse_skin/multiome_ncells_pmat/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [5000]\n",
    "repeats = 5\n",
    "output_folder=\"results_single_same_cell_number\"\n",
    "\n",
    "\n",
    "for i in range(len(iter_list)):\n",
    "    for j in range(1,repeats+1):\n",
    "        in_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j))\n",
    "        out_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j),output_folder)\n",
    "\n",
    "        os.makedirs(os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\"), exist_ok=True)\n",
    "        # save ATAC to the right location\n",
    "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"), \n",
    "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"barcodes.tsv\"))\n",
    "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"peak.tsv\"), \n",
    "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"peak.tsv\"))\n",
    "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"ATAC_counts.mtx\"), \n",
    "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"counts.mtx\"))\n",
    "\n",
    "        # load in ATAC barcodes \n",
    "        atac_bc = pd.read_csv(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"),header=None)\n",
    "\n",
    "        # get RNA profile of the corresponding cells \n",
    "        in_dir = \"dataset/mouse_skin/\"\n",
    "        adata_rna = ad.read_h5ad(in_dir+\"mouse_skin_shareseq_rna.h5ad\")\n",
    "\n",
    "        idx_sel = atac_bc[0].tolist()\n",
    "        # RNA\n",
    "        adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna[idx_sel,:].X.todense())),\n",
    "                               obs=adata_rna.obs.loc[idx_sel,:],\n",
    "                               var=adata_rna.var,dtype=np.float32)\n",
    "        adata_rna_sel.obs['rna.bc'] = adata_rna_sel.obs_names\n",
    "        adata_rna_sel.var['feature'] = deepcopy(adata_rna_sel.var_names)\n",
    "\n",
    "        # save RNA to the right location\n",
    "        utils_eval.write_adata(adata_rna_sel, os.path.join(out_dir_i,\"truth\",\"predicted\",\"RNA\"),\"RNA\",\"gene\",feature_name='feature',transpose=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run eval_missming_modality_prediction \n",
    "\n",
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_missing_mod_eval.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_single.py\"\n",
    "\n",
    "ct_ref = \"dataset/mouse_skin/mouse_skin_shareseq_bc_ct3.csv\"\n",
    "nclust = 22\n",
    "\n",
    "dir_path = \"dataset/mouse_skin/multiome_ncells_pmat/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [5000]\n",
    "repeats = 5\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single_mm10.R\"\n",
    "gp_truth = \"dataset/mouse_skin/mouse_skin_shareseq_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
    "\n",
    "method_keys = [\"truth\"]\n",
    "\n",
    "#----- All of these can be random strings, are not used in script, but need not to be empty strings ----\n",
    "conda_envs = [\"truth_na\"]\n",
    "# R script for running the method\n",
    "method_scripts = [\"run_truth.csv\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "file_paths = [\"truth/truth_result.csv\"]\n",
    "#------ END ----- \n",
    "\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=2,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 2 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PBMC 2000 cells "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_dir = \"dataset/multiome_pbmc_10k/\"\n",
    "\n",
    "adata_rna = ad.read_h5ad(source_dir+\"pbmc_10x_rna_public.h5ad\")\n",
    "adata_atac = ad.read_h5ad(source_dir+\"pbmc_10x_atac_public.h5ad\")\n",
    "\n",
    "\n",
    "in_dir = \"dataset/multiome_pbmc_10k/nmulti2000_7ct_vdepth_test/\"\n",
    "iter_list = [0.25,0.5,0.75,1]\n",
    "depth_multiome_list = iter_list\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = [2000]*len(iter_list)\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "fragment_path = \"{}/dataset/multiome_pbmc_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "\n",
    "cond_key = \"depthmulti\"\n",
    "\n",
    "def str_norm(s):\n",
    "    return(str(int(s*100)))\n",
    "\n",
    "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
    "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                repeats,fragment_path,cond_key,\n",
    "                str_norm, downsample=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_fair.py\"\n",
    "\n",
    "ct_ref = \"dataset/multiome_pbmc_10k/pbmc_10x_bc_ct3.csv\"\n",
    "nclust = 7\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
    "gp_truth = \"dataset/multiome_pbmc_10k/pbmc_10x_pmat_sig_links_50kb_unique.csv\"\n",
    "\n",
    "dir_path = \"dataset/multiome_pbmc_10k/nmulti2000_7ct_vdepth_test/\"\n",
    "cond_key = \"depthmulti\"\n",
    "\n",
    "iter_list = [25,50,75,100]\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*2 \n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "             \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# Two additional methods\n",
    "conda_envs = [\"glue2\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38.py\",\"run_scmomat_pbmc.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*2\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\",\"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\",\"scmomat\"]\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=1,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BMMC 2000 cells "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using pmat dataset, generated from bmmc_process R script \n",
    "in_dir = \"dataset/bmmc/\"\n",
    "\n",
    "adata_atac = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
    "adata_rna = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "adata_atac.var['feature'] = adata_atac.var['features'] \n",
    "adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "# aggregate the selected donors \n",
    "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
    "\n",
    "# Paired - RNA\n",
    "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "# Paired - ATAC\n",
    "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "\n",
    "# ensure that var_names are feature names and obs_names are cell barcodes\n",
    "adata_rna_sel.var_names = adata_rna.var['feature'].tolist()\n",
    "adata_atac_sel.var_names = adata_atac.var['feature'].tolist()\n",
    "\n",
    "in_dir = \"dataset/bmmc/nmulti2000_21ct_vdepth_test/\"\n",
    "iter_list = [0.25]#[0.25,0.5,0.75,1]\n",
    "depth_multiome_list = iter_list\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = [2000]*len(iter_list)\n",
    "repeats = 2 #5\n",
    "\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
    "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "cond_key = \"depthmulti\"\n",
    "\n",
    "def str_norm(s):\n",
    "    return(str(int(s*100)))\n",
    "\n",
    "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
    "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                repeats,fragment_path,cond_key,\n",
    "                str_norm, downsample=True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_fair.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
    "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/nmulti2000_21ct_vdepth_test/\"\n",
    "cond_key = \"depthmulti\"\n",
    "\n",
    "iter_list = [25,50,75,100]\n",
    "repeats = 5\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*2 \n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "             \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# Two additional methods\n",
    "conda_envs = [\"glue2\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38.py\",\"run_scmomat_bmmc.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*2\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\",\"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\",\"scmomat\"]\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=1,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BMMC 4000 cells "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using pmat dataset, generated from bmmc_process R script \n",
    "in_dir = \"dataset/bmmc/\"\n",
    "\n",
    "adata_atac = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
    "adata_rna = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "adata_atac.var['feature'] = adata_atac.var['features'] \n",
    "adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "# aggregate the selected donors \n",
    "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
    "\n",
    "# Paired - RNA\n",
    "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "# Paired - ATAC\n",
    "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "\n",
    "# ensure that var_names are feature names and obs_names are cell barcodes\n",
    "adata_rna_sel.var_names = adata_rna.var['feature'].tolist()\n",
    "adata_atac_sel.var_names = adata_atac.var['feature'].tolist()\n",
    "\n",
    "in_dir = \"dataset/bmmc/nmulti4000_21ct_vdepth_test/\"\n",
    "iter_list = [0.25,0.5,0.75,1]\n",
    "depth_multiome_list = iter_list\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = [4000]*len(iter_list)\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
    "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "cond_key = \"depthmulti\"\n",
    "\n",
    "def str_norm(s):\n",
    "    return(str(int(s*100)))\n",
    "\n",
    "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
    "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                repeats,fragment_path,cond_key,\n",
    "                str_norm, downsample=True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_fair.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
    "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/nmulti4000_21ct_vdepth_test/\"\n",
    "cond_key = \"depthmulti\"\n",
    "\n",
    "iter_list = [25,50,75,100]\n",
    "repeats = 5\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*2 \n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "             \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# Two additional methods\n",
    "conda_envs = [\"glue2\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38.py\",\"run_scmomat_bmmc.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*2\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\",\"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\",\"scmomat\"]\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=1,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BMMC Increasing number of cells; 100% depth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_dir = \"dataset/bmmc/\"\n",
    "\n",
    "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
    "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "adata_atac.var['feature'] = adata_atac.var['features'] \n",
    "adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "\n",
    "# aggregate the selected donors \n",
    "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
    "\n",
    "# Paired - RNA\n",
    "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "# Paired - ATAC\n",
    "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "\n",
    "# ensure that var_names are feature names and obs_names are cell barcodes\n",
    "adata_rna_sel.var_names = adata_rna.var['feature'].tolist()\n",
    "adata_atac_sel.var_names = adata_atac.var['feature'].tolist()\n",
    "\n",
    "\n",
    "in_dir = \"dataset/bmmc/bmmc_vcells_intervals/\"\n",
    "iter_list = [int(x) for x in np.linspace(start=1000, stop=1000+400*10, num=11)] # ncell increase interval, each interval: increase by 400 cells. 400 cells of 10k atac fragment and 2500 nCounts_RNA is roughly equal to 10% increment in ATAC and RNA sequencing depth, assuming 4000 cells\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "repeats = 5\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
    "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "cond_key = \"nmulti\"\n",
    "    \n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "\n",
    "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
    "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                repeats,fragment_path,cond_key,\n",
    "                to_str, downsample=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_fair.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
    "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/bmmc_vcells_intervals/\"\n",
    "cond_key = \"depthmulti\"\n",
    "\n",
    "iter_list = [int(x) for x in np.linspace(start=1000, stop=1000+400*10, num=11)] # ncell increase interval, each interval: increase by 400 cells. 400 cells of 10k atac fragment and 2500 nCounts_RNA is roughly equal to 10% increment in ATAC and RNA sequencing depth, assuming 4000 cells\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# ===== Seurat v4 ====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== Seurat v3 ===== \n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BMMC Increasing depth; 4000 cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_dir = \"dataset/bmmc/\"\n",
    "\n",
    "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
    "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "adata_atac.var['feature'] = adata_atac.var['features'] \n",
    "adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "# aggregate the selected donors \n",
    "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
    "\n",
    "# Paired - RNA\n",
    "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "# Paired - ATAC\n",
    "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "\n",
    "# ensure that var_names are feature names and obs_names are cell barcodes\n",
    "adata_rna_sel.var_names = adata_rna.var['feature'].tolist()\n",
    "adata_atac_sel.var_names = adata_atac.var['feature'].tolist()\n",
    "\n",
    "\n",
    "# Saving everything in /project/mingyaolpc/ folder to avoid running out of storage\n",
    "in_dir = \"dataset/bmmc/bmmc_vdepth_intervals/\"\n",
    "iter_list = np.linspace(start=10, stop=100, num=10)/100\n",
    "depth_multiome_list = iter_list\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = [4000]*len(iter_list)\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
    "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "cond_key = \"depthmulti\"\n",
    "\n",
    "def str_norm(s):\n",
    "    return(str(int(s*100)))\n",
    "\n",
    "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
    "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                repeats,fragment_path,cond_key,\n",
    "                str_norm, downsample=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_fair.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\n",
    "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
    "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/bmmc_vdepth_intervals/\"\n",
    "cond_key = \"depthmulti\"\n",
    "\n",
    "iter_list = np.linspace(start=10, stop=100, num=10)/100\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# ===== Seurat v4 =====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== Seurat v3 ===== \n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Technical batch effect "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using pmat dataset, generated from bmmc_process R script \n",
    "source_dir = \"dataset/bmmc/\"\n",
    "\n",
    "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
    "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "adata_atac.var['feature'] = adata_atac.var['features'] \n",
    "adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "# Paired\n",
    "idx_s1d1 = list(adata_atac.obs['batch'] == 's1d1')\n",
    "\n",
    "# Paired - RNA\n",
    "adata_rna_s1d1 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s1d1,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_s1d1,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "# Paired - ATAC\n",
    "adata_atac_s1d1 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s1d1,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_s1d1,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "# snRNA\n",
    "idx_s2d1 = list(adata_atac.obs['batch'] == 's2d1')\n",
    "adata_rna_s2d1 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s2d1,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_s2d1,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "\n",
    "# snATAC\n",
    "adata_atac_s2d1 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s2d1,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_s2d1,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "\n",
    "in_dir = \"dataset/bmmc/bmmc_technical_batch_test/\"\n",
    "iter_list = [1000,3000,5000]\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "repeats = 5\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
    "fragment_path = \"{}/dataset/bmmc/bmmc_site1_or_donor1_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "cond_key = \"nmulti\"\n",
    "    \n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "# assume there are ['rna.bc','batch'] in each adata.obs\n",
    "# assume there is ['feature'] in each adata.var\n",
    "data_simulation_batch(in_dir,adata_rna_s1d1,adata_atac_s1d1,adata_rna_s2d1,adata_atac_s2d1,\n",
    "                      iter_list,depth_multiome_list,depth_scrna_list,\n",
    "                      depth_snatac_list,n_multiome_list,\n",
    "                      n_scrna_list,n_snatac_list,\n",
    "                      repeats,fragment_path,cond_key,\n",
    "                      to_str, downsample=False,same_unpair_origin=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_batch2.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\\\n",
    "gp_eval_path = \"false\"\n",
    "gp_truth = \"false\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/bmmc_technical_batch_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [1000,3000,5000]\n",
    "repeats = 5\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_batch.py\",\"run_cobolt.py\",\"run_scmomat_batch_bmmc_2.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*3\n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "             \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# Additional python method\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38_batch.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Biological batch effect "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using pmat dataset, generated from bmmc_process R script \n",
    "source_dir = \"dataset/bmmc/\"\n",
    "\n",
    "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
    "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "adata_atac.var['feature'] = adata_atac.var['features'] \n",
    "adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "# Paired\n",
    "idx_s1d1 = list(adata_atac.obs['batch'] == 's1d1')\n",
    "\n",
    "# Paired - RNA\n",
    "adata_rna_s1d1 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s1d1,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_s1d1,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "# Paired - ATAC\n",
    "adata_atac_s1d1 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s1d1,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_s1d1,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "# snRNA\n",
    "idx_s1d2 = list(adata_atac.obs['batch'] == 's1d2')\n",
    "adata_rna_s1d2 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s1d2,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_s1d2,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "\n",
    "# snATAC\n",
    "adata_atac_s1d2 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s1d2,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_s1d2,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "\n",
    "in_dir = \"dataset/bmmc/bmmc_biological_batch_test/\"\n",
    "iter_list = [1000,3000,5000]\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "repeats = 5\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
    "fragment_path = \"{}/dataset/bmmc/bmmc_site1_or_donor1_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "cond_key = \"nmulti\"\n",
    "    \n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "# assume there are ['rna.bc','batch'] in each adata.obs\n",
    "# assume there is ['feature'] in each adata.var\n",
    "data_simulation_batch(in_dir,adata_rna_s1d1,adata_atac_s1d1,adata_rna_s2d1,adata_atac_s2d1,\n",
    "                      iter_list,depth_multiome_list,depth_scrna_list,\n",
    "                      depth_snatac_list,n_multiome_list,\n",
    "                      n_scrna_list,n_snatac_list,\n",
    "                      repeats,fragment_path,cond_key,\n",
    "                      to_str, downsample=False,same_unpair_origin=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_batch2.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\n",
    "gp_eval_path = \"false\"\n",
    "gp_truth = \"false\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/bmmc_biological_batch_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [1000,3000,5000]\n",
    "repeats = 5\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_batch.py\",\"run_cobolt.py\",\"run_scmomat_batch_bmmc_2.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*3\n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "             \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# Additional python method\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38_batch.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Complex test #1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using pmat dataset, generated from bmmc_process R script \n",
    "source_dir = \"dataset/bmmc/\"\n",
    "\n",
    "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
    "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "adata_atac.var['feature'] = adata_atac.var['features'] \n",
    "adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "# Paired\n",
    "idx_paired = list(adata_atac.obs['batch'].isin(['s1d1','s1d3']))\n",
    "\n",
    "# Paired - RNA\n",
    "adata_rna_paired = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_paired,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_paired,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "# Paired - ATAC\n",
    "adata_atac_paired = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_paired,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_paired,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "# snRNA\n",
    "idx_s1d2 = list(adata_atac.obs['batch'] == 's1d2')\n",
    "adata_rna_s1d2 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s1d2,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_s1d2,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "\n",
    "# snATAC\n",
    "adata_atac_s1d2 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s1d2,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_s1d2,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "\n",
    "in_dir = \"dataset/bmmc/bmmc_complex1_test/\"\n",
    "iter_list = [1000,3000,5000]\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "repeats = 5\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
    "fragment_path = \"{}/dataset/bmmc/bmmc_site1_or_donor1_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "cond_key = \"nmulti\"\n",
    "    \n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "# assume there are ['rna.bc','batch'] in each adata.obs\n",
    "# assume there is ['feature'] in each adata.var\n",
    "data_simulation_batch(in_dir,adata_rna_s1d1,adata_atac_s1d1,adata_rna_s2d1,adata_atac_s2d1,\n",
    "                      iter_list,depth_multiome_list,depth_scrna_list,\n",
    "                      depth_snatac_list,n_multiome_list,\n",
    "                      n_scrna_list,n_snatac_list,\n",
    "                      repeats,fragment_path,cond_key,\n",
    "                      to_str, downsample=False,same_unpair_origin=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_batch2.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\n",
    "gp_eval_path = \"false\"\n",
    "gp_truth = \"false\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/bmmc_complex1_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [1000,3000,5000]\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_batch.py\",\"run_cobolt.py\",\"run_scmomat_batch_bmmc_2.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*3\n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\",\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\",\"run_seurat4_4.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*2\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\",\"seurat4int/seurat4int_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4int\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "             \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# Additional python method\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38_batch.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Complex test #2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using pmat dataset, generated from bmmc_process R script \n",
    "source_dir = \"dataset/bmmc/\"\n",
    "\n",
    "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
    "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
    "\n",
    "adata_atac.var['feature'] = adata_atac.var['features'] \n",
    "adata_rna.var['feature'] = adata_rna.var['features'] \n",
    "\n",
    "# Paired\n",
    "idx_paired = list(adata_atac.obs['batch'].isin(['s4d1','s1d3']))\n",
    "\n",
    "# Paired - RNA\n",
    "adata_rna_paired = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_paired,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_paired,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "# Paired - ATAC\n",
    "adata_atac_paired = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_paired,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_paired,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "# snRNA\n",
    "idx_s2d1 = list(adata_atac.obs['batch'] == 's2d1')\n",
    "adata_rna_s2d1 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s2d1,:].todense())),\n",
    "                       obs=adata_rna.obs.iloc[idx_s2d1,:],\n",
    "                       var=adata_rna.var,dtype=np.float32)\n",
    "\n",
    "# snATAC\n",
    "adata_atac_s2d1 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s2d1,:].todense())),\n",
    "                       obs=adata_atac.obs.iloc[idx_s2d1,:],\n",
    "                       var=adata_atac.var,dtype=np.float32)\n",
    "\n",
    "in_dir = \"dataset/bmmc/bmmc_complex2_test/\"\n",
    "iter_list = [10000]\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list) \n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "n_scrna_list  = [2000]*len(iter_list)\n",
    "n_snatac_list = [2000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
    "fragment_path = \"{}/dataset/bmmc/bmmc_site1_or_donor1_atac_fragments.tsv.gz\".format(os.getcwd())\n",
    "cond_key = \"nmulti\"\n",
    "    \n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "# assume there are ['rna.bc','batch'] in each adata.obs\n",
    "# assume there is ['feature'] in each adata.var\n",
    "data_simulation_batch(in_dir,adata_rna_s1d1,adata_atac_s1d1,adata_rna_s2d1,adata_atac_s2d1,\n",
    "                      iter_list,depth_multiome_list,depth_scrna_list,\n",
    "                      depth_snatac_list,n_multiome_list,\n",
    "                      n_scrna_list,n_snatac_list,\n",
    "                      repeats,fragment_path,cond_key,\n",
    "                      to_str, downsample=False,same_unpair_origin=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"run_metric_eval_batch2.py\"\n",
    "\n",
    "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
    "nclust = 21\n",
    "\n",
    "gp_eval_path = \"false\"\n",
    "gp_truth = \"false\"\n",
    "\n",
    "dir_path = \"dataset/bmmc/bmmc_complex2_test/\"\n",
    "cond_key = \"nmulti\"\n",
    "\n",
    "iter_list = [10000]\n",
    "repeats = 5\n",
    "\n",
    "# ===== multiome-guided (python functions) =====\n",
    "\n",
    "conda_envs = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_batch.py\",\"run_cobolt.py\",\"run_scmomat_batch_bmmc_2.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*3\n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"cobolt\",\"scmomat\"]\n",
    "\n",
    "\n",
    "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== multiome-guided (R functions) =====\n",
    "\n",
    "conda_envs = [\"seurat\",\"seurat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat4_3.R\",\"run_seurat4_4.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*2\n",
    "\n",
    "file_paths = [\"seurat4/seurat4_result.csv\",\"seurat4int/seurat4int_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat4int\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "\n",
    "# ===== unpaired (multiome-split) ===== \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "             \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "# submit all jobs at once\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n",
    "# Additional python method\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38_batch.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PBMC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_dir = \"dataset/multiome_pbmc_10k/\"\n",
    "\n",
    "adata_rna_sel = ad.read_h5ad(source_dir+\"pbmc_10x_rna_public.h5ad\")\n",
    "adata_atac_sel = ad.read_h5ad(source_dir+\"pbmc_10x_atac_public.h5ad\")\n",
    "\n",
    "adata_rna_sel.obs['ct3'].value_counts()\n",
    "\n",
    "in_dir = \"dataset/pbmc/single_modality_fixed_missing_ct/\"\n",
    "\n",
    "iter_list = [1000,3000,6000]\n",
    "cts_remove_multiome_list = [[None],[None],[None]]\n",
    "cts_remove_scrna_list = [[None],['NK'],[None]]\n",
    "cts_remove_snatac_list = [[None],[None],['NK']]\n",
    "\n",
    "percent_single_mod_ct = {'NK': 0.046}\n",
    "\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list)\n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "\n",
    "n_scrna_list = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "fragment_path = \"dataset/multiome_pbmc_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz\"\n",
    "\n",
    "cond_keys = [\"noMiss_nmulti\",\"rnaMissNK_nmulti\",\"atacMissNK_nmulti\"]\n",
    "\n",
    "# the obs column storing cell type annotation \n",
    "ct_col = 'ct3'\n",
    "\n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "from data_simulation import simulate_missing_fixed\n",
    "for i in range(len(cond_keys)):\n",
    "    simulate_missing_fixed(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
    "                           depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                           n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                           # for each cond_key, have a different ct structure. Each time, create nmulti cells in a situation\n",
    "                           cts_remove_multiome_list[i], \n",
    "                           cts_remove_scrna_list[i],\n",
    "                           cts_remove_snatac_list[i],\n",
    "                           percent_single_mod_ct,repeats,fragment_path,\n",
    "                           cond_keys[i],\n",
    "                           to_str, downsample=False,ct_col=ct_col)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_missing_ct_eval.sh\"\n",
    "\n",
    "# python script for metric evaluation \n",
    "# this one evaluation script works for single-modality focused analysis and all three datasets analysis\n",
    "# Because F1 score is calculated per cell type per modality\n",
    "eval_script = \"run_metric_eval_fair_missing_ct_perMod.py\"\n",
    "\n",
    "# write the rare_ct_list to folder, so the evaluation part knows to calculate F1 score for these two cell types\n",
    "dir_path = \"dataset/pbmc/single_modality_fixed_missing_ct/\"\n",
    "\n",
    "rare_ct_path = os.path.join(dir_path,\"rare_ct_list.csv\")\n",
    "rare_ct_list = ['NK']\n",
    "pd.DataFrame(rare_ct_list).to_csv(rare_ct_path)\n",
    "\n",
    "ct_ref = \"dataset/multiome_pbmc_10k/pbmc_10x_bc_ct3.csv\"\n",
    "nclust = 7\n",
    "\n",
    "gp_eval_path = \"false\"\n",
    "gp_truth = \"false\"\n",
    "\n",
    "cond_keys = [\"noMiss_nmulti\",\"rnaMissNK_nmulti\",\"atacMissNK_nmulti\"]\n",
    "\n",
    "\n",
    "iter_list = [1000,3000,6000]\n",
    "repeats = 5\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Unpaired + seurat 4\n",
    "\n",
    "# all R functions \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"seurat\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3_single_noPred.R\",\"run_rbindsc_single_noPred.R\",\n",
    "                  \"run_seurat4_3_noPred.R\",\"run_rfigr_single.R\",\"run_rliger_single.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*5\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"seurat4/seurat4_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"seurat4\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=False, wait_time=1*30,\n",
    "                  batch=1,output_folder=\"results_single_mod\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "### Unpaired (multiome-split)\n",
    "\n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts =  [\"run_seurat3_noPred.R\",\"run_rbindsc_noPred.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=False, wait_time=1*30,\n",
    "                  batch=1,output_folder=\"results_single_same_cell_number\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "### python functions\n",
    "conda_envs = [\"scmomat\",\"cobolt\",\"glue2\",\"multivi\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_scmomat_pbmc.py\",\n",
    "                  \"run_cobolt.py\",\n",
    "                  \"run_glue_hg38_single_noPred.py\",\n",
    "                  \"run_multivi_2_noPred.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*4\n",
    "\n",
    "file_paths = [\"scmomat/scmomat_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"glue/glue_result.csv\",\n",
    "              \"multivi/multivi_result.csv\"]\n",
    "\n",
    "method_keys = [\"scmomat\",\"cobolt\",\"glue\",\"multivi\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=True, wait_time=1*30,\n",
    "                  batch=4,output_folder=\"results_single_mod\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38_noPred.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=True, wait_time=1*30,\n",
    "                  batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SHARE-seq "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load \n",
    "source_dir = \"dataset/mouse_skin/\"\n",
    "adata_rna = ad.read_h5ad(source_dir+\"mouse_skin_shareseq_rna_10k.h5ad\")\n",
    "adata_atac = ad.read_h5ad(source_dir+\"mouse_skin_shareseq_atac_10k.h5ad\")\n",
    "\n",
    "\n",
    "in_dir = \"dataset/mouse_skin/single_modality_fixed_missing_ct/\"\n",
    "\n",
    "iter_list = [1000,3000,6000]\n",
    "cts_remove_multiome_list = [[None],[None],[None],[None],[None],[None],[None],[None],[None]]\n",
    "cts_remove_scrna_list = [[None],['HS'],[None],['Endo'],[None],['HS','Endo'],[None],['HS'],['Endo']]\n",
    "cts_remove_snatac_list = [[None],[None],['HS'],[None],['Endo'],[None],['HS','Endo'],['Endo'],['HS']]\n",
    "\n",
    "percent_single_mod_ct = {'HS': 0.1, 'Endo':0.05}\n",
    "\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list)\n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "\n",
    "n_scrna_list = [1000]*len(iter_list)\n",
    "n_snatac_list = [1000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "fragment_path = \"dataset/mouse_skin/mouse_skin_shareseq_fragments.tsv.gz\"\n",
    "\n",
    "cond_keys = [\"noMiss_nmulti\",\"rnaMissHS_nmulti\",\"atacMissHS_nmulti\",\"rnaMissEndo_nmulti\",\"atacMissEndo_nmulti\",\n",
    "            \"rnaMissTwo_nmulti\",\"atacMissTwo_nmulti\",\"eachMissOne_nmulti\",\"eachMissOneAlt_nmulti\"]\n",
    "\n",
    "# the obs column storing cell type annotation \n",
    "ct_col = 'ct3'\n",
    "\n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "from data_simulation import simulate_missing_fixed\n",
    "for i in range(len(cond_keys)):\n",
    "    simulate_missing_fixed(in_dir,adata_rna,adata_atac,iter_list,\n",
    "                           depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                           n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                           # for each cond_key, have a different ct structure. Each time, create nmulti cells in a situation\n",
    "                           cts_remove_multiome_list[i], \n",
    "                           cts_remove_scrna_list[i],\n",
    "                           cts_remove_snatac_list[i],\n",
    "                           percent_single_mod_ct,repeats,fragment_path,\n",
    "                           cond_keys[i],\n",
    "                           to_str, downsample=False,ct_col=ct_col)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_missing_ct_eval.sh\"\n",
    "\n",
    "# python script for metric evaluation \n",
    "# this one evaluation script works for single-modality focused analysis and all three datasets analysis\n",
    "# Because F1 score is calculated per cell type per modality\n",
    "eval_script = \"run_metric_eval_fair_missing_ct_perMod.py\"\n",
    "\n",
    "# write the rare_ct_list to folder, so the evaluation part knows to calculate F1 score for these two cell types\n",
    "dir_path = \"dataset/mouse_skin/single_modality_fixed_missing_ct/\"\n",
    "\n",
    "rare_ct_path = os.path.join(dir_path,\"rare_ct_list.csv\")\n",
    "rare_ct_list = ['HS','Endo']\n",
    "pd.DataFrame(rare_ct_list).to_csv(rare_ct_path)\n",
    "\n",
    "ct_ref = \"dataset/mouse_skin/mouse_skin_shareseq_bc_ct3_10k.csv\"\n",
    "nclust = 12\n",
    "\n",
    "gp_eval_path = \"false\"\n",
    "gp_truth = \"false\"\n",
    "\n",
    "cond_keys = [\"noMiss_nmulti\",\"rnaMissHS_nmulti\",\"atacMissHS_nmulti\",\"rnaMissEndo_nmulti\",\"atacMissEndo_nmulti\",\n",
    "            \"rnaMissTwo_nmulti\",\"atacMissTwo_nmulti\",\"eachMissOne_nmulti\",\"eachMissOneAlt_nmulti\"]\n",
    "\n",
    "iter_list = [1000,3000,6000]\n",
    "repeats = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Unpaired + seurat 4\n",
    "\n",
    "# all R functions \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"seurat\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3_single_noPred.R\",\n",
    "                  \"run_rbindsc_single_noPred.R\",\n",
    "                  \"run_seurat4_3_noPred.R\",\n",
    "                  \"run_rfigr_single.R\",\n",
    "                  \"run_rliger_single.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*5\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"seurat4/seurat4_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"seurat4\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=False, wait_time=1*30,\n",
    "                  batch=1,output_folder=\"results_single_mod\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "### Unpaired (multiome-split)\n",
    "\n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts =  [\"run_seurat3_noPred.R\",\n",
    "                   \"run_rbindsc_noPred.R\",\n",
    "                   \"run_rfigr_2.R\",\n",
    "                   \"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=False, wait_time=1*30,\n",
    "                  batch=1,output_folder=\"results_single_same_cell_number\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "### python functions\n",
    "conda_envs = [\"scmomat\",\"cobolt\",\"glue2\",\"multivi\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_scmomat_mouse_skin.py\",\n",
    "                  \"run_cobolt.py\",\n",
    "                  \"run_glue_mm10_single_noPred.py\",\n",
    "                  \"run_multivi_2_noPred.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*4\n",
    "\n",
    "file_paths = [\"scmomat/scmomat_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"glue/glue_result.csv\",\n",
    "              \"multivi/multivi_result.csv\"]\n",
    "\n",
    "method_keys = [\"scmomat\",\"cobolt\",\"glue\",\"multivi\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=True, wait_time=1*30,\n",
    "                  batch=4,output_folder=\"results_single_mod\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_mm10_noPred.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=True, wait_time=1*30,\n",
    "                  batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scenario 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PBMC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_dir = \"dataset/multiome_pbmc_10k/\"\n",
    "\n",
    "adata_rna_sel = ad.read_h5ad(source_dir+\"pbmc_10x_rna_public.h5ad\")\n",
    "adata_atac_sel = ad.read_h5ad(source_dir+\"pbmc_10x_atac_public.h5ad\")\n",
    "\n",
    "adata_rna_sel.obs['ct3'].value_counts()\n",
    "\n",
    "in_dir = \"dataset/pbmc/multiome_fixed_missing_ct/\"\n",
    "\n",
    "iter_list = [3000]\n",
    "cts_remove_multiome_list = [[None],['NK'],['NK'],['NK']]\n",
    "cts_remove_scrna_list =    [[None],[None],[None],['NK']]\n",
    "cts_remove_snatac_list =   [[None],[None],['NK'],[None]]\n",
    "\n",
    "percent_single_mod_ct = {'NK': 0.046}\n",
    "\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list)\n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "\n",
    "n_scrna_list = [3000]*len(iter_list)\n",
    "n_snatac_list = [3000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "\n",
    "repeats = 5\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "fragment_path = \"dataset/multiome_pbmc_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz\"\n",
    "\n",
    "cond_keys = [\"noMiss_nmulti\",\"multiMissNK_nmulti\",\"rnaOnlyNK_nmulti\",\"atacOnlyNK_nmulti\"]\n",
    "\n",
    "# the obs column storing cell type annotation \n",
    "ct_col = 'ct3'\n",
    "\n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "from data_simulation import simulate_missing_fixed\n",
    "for i in range(len(cond_keys)):\n",
    "    simulate_missing_fixed(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
    "                           depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                           n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                           # for each cond_key, have a different ct structure. Each time, create nmulti cells in a situation\n",
    "                           cts_remove_multiome_list[i], \n",
    "                           cts_remove_scrna_list[i],\n",
    "                           cts_remove_snatac_list[i],\n",
    "                           percent_single_mod_ct,repeats,fragment_path,\n",
    "                           cond_keys[i],\n",
    "                           to_str, downsample=False,ct_col=ct_col)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_missing_ct_eval.sh\"\n",
    "\n",
    "# python script for metric evaluation \n",
    "# this one evaluation script works for single-modality focused analysis and all three datasets analysis\n",
    "# Because F1 score is calculated per cell type per modality\n",
    "eval_script = \"run_metric_eval_fair_missing_ct_perMod.py\"\n",
    "\n",
    "# write the rare_ct_list to folder, so the evaluation part knows to calculate F1 score for these two cell types\n",
    "dir_path = \"dataset/pbmc/multiome_fixed_missing_ct/\"\n",
    "\n",
    "rare_ct_path = os.path.join(dir_path,\"rare_ct_list.csv\")\n",
    "rare_ct_list = ['NK']\n",
    "pd.DataFrame(rare_ct_list).to_csv(rare_ct_path)\n",
    "\n",
    "ct_ref = \"dataset/multiome_pbmc_10k/pbmc_10x_bc_ct3.csv\"\n",
    "nclust = 7\n",
    "\n",
    "gp_eval_path = \"false\"\n",
    "gp_truth = \"false\"\n",
    "\n",
    "cond_keys = [\"noMiss_nmulti\",\"multiMissNK_nmulti\",\"rnaOnlyNK_nmulti\",\"atacOnlyNK_nmulti\"]\n",
    "iter_list = [3000]\n",
    "repeats = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Unpaired + seurat 4\n",
    "\n",
    "# all R functions \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"seurat\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3_single_noPred.R\",\n",
    "                  \"run_rbindsc_single_noPred.R\",\n",
    "                  \"run_seurat4_3_noPred.R\",\n",
    "                  \"run_rfigr_single.R\",\n",
    "                  \"run_rliger_single.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*5\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"seurat4/seurat4_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"seurat4\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=False, wait_time=1*30,\n",
    "                  batch=1,output_folder=\"results_single_mod\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "### Unpaired (multiome-split)\n",
    "\n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts =  [\"run_seurat3_noPred.R\",\"run_rbindsc_noPred.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=False, wait_time=1*30,\n",
    "                  batch=1,output_folder=\"results_single_same_cell_number\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "### python functions\n",
    "conda_envs = [\"scmomat\",\"cobolt\",\"glue2\",\"multivi\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_scmomat_pbmc.py\",\n",
    "                  \"run_cobolt.py\",\n",
    "                  \"run_glue_hg38_single_noPred.py\",\n",
    "                  \"run_multivi_2_noPred.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*4\n",
    "\n",
    "file_paths = [\"scmomat/scmomat_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"glue/glue_result.csv\",\n",
    "              \"multivi/multivi_result.csv\"]\n",
    "\n",
    "method_keys = [\"scmomat\",\"cobolt\",\"glue\",\"multivi\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=True, wait_time=1*30,\n",
    "                  batch=4,output_folder=\"results_single_mod\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_hg38_noPred.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=True, wait_time=1*30,\n",
    "                  batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SHARE-seq "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load \n",
    "source_dir = \"dataset/mouse_skin/\"\n",
    "adata_rna = ad.read_h5ad(source_dir+\"mouse_skin_shareseq_rna_10k.h5ad\")\n",
    "adata_atac = ad.read_h5ad(source_dir+\"mouse_skin_shareseq_atac_10k.h5ad\")\n",
    "\n",
    "in_dir = \"dataset/mouse_skin/multiome_fixed_missing_ct/\"\n",
    "\n",
    "iter_list = [3000]\n",
    "cts_remove_multiome_list = [[None],['HS'],['HS'],['HS'],['Endo'],['Endo'],['Endo']]\n",
    "cts_remove_scrna_list =    [[None],[None],[None],['HS'],[None],[None],['Endo']]\n",
    "cts_remove_snatac_list =   [[None],[None],['HS'],[None],[None],['Endo'],[None]]\n",
    "\n",
    "percent_single_mod_ct = {'HS': 0.1, 'Endo':0.05}\n",
    "\n",
    "depth_multiome_list = [1]*len(iter_list)\n",
    "depth_scrna_list = [1]*len(iter_list)\n",
    "depth_snatac_list = [1]*len(iter_list)\n",
    "\n",
    "n_scrna_list = [3000]*len(iter_list)\n",
    "n_snatac_list = [3000]*len(iter_list)\n",
    "n_multiome_list = iter_list\n",
    "\n",
    "repeats = 5\n",
    "\n",
    "\n",
    "# if do not create fragment file symlink, use fragment_path = None \n",
    "fragment_path = \"dataset/mouse_skin/mouse_skin_shareseq_fragments.tsv.gz\"\n",
    "\n",
    "#cond_key = \"nmulti\"\n",
    "cond_keys = [\"noMiss_nmulti\",\"multiMissHS_nmulti\",\"rnaOnlyHS_nmulti\",\"atacOnlyHS_nmulti\",\n",
    "             \"multiMissEndo_nmulti\",\"rnaOnlyEndo_nmulti\",\"atacOnlyEndo_nmulti\"]\n",
    "\n",
    "# the obs column storing cell type annotation \n",
    "ct_col = 'ct3'\n",
    "\n",
    "def to_str(s): return(str(s))\n",
    "\n",
    "from data_simulation import simulate_missing_fixed\n",
    "for i in range(len(cond_keys)):\n",
    "    simulate_missing_fixed(in_dir,adata_rna,adata_atac,iter_list,\n",
    "                           depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
    "                           n_multiome_list,n_scrna_list,n_snatac_list,\n",
    "                           # for each cond_key, have a different ct structure. Each time, create nmulti cells in a situation\n",
    "                           cts_remove_multiome_list[i], \n",
    "                           cts_remove_scrna_list[i],\n",
    "                           cts_remove_snatac_list[i],\n",
    "                           percent_single_mod_ct,repeats,fragment_path,\n",
    "                           cond_keys[i],\n",
    "                           to_str, downsample=False,ct_col=ct_col)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running method + evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "\n",
    "# bash script for job submission\n",
    "job_submission_script = \"submit_job_per_condition_n_missing_ct_eval.sh\"\n",
    "\n",
    "# python script for metric evaluation \n",
    "# this one evaluation script works for single-modality focused analysis and all three datasets analysis\n",
    "# Because F1 score is calculated per cell type per modality\n",
    "eval_script = \"run_metric_eval_fair_missing_ct_perMod.py\"\n",
    "\n",
    "# write the rare_ct_list to folder, so the evaluation part knows to calculate F1 score for these two cell types\n",
    "dir_path = \"dataset/mouse_skin/multiome_fixed_missing_ct/\"\n",
    "\n",
    "rare_ct_path = os.path.join(dir_path,\"rare_ct_list.csv\")\n",
    "rare_ct_list = ['HS','Endo']\n",
    "pd.DataFrame(rare_ct_list).to_csv(rare_ct_path)\n",
    "\n",
    "ct_ref = \"dataset/mouse_skin/mouse_skin_shareseq_bc_ct3_10k.csv\"\n",
    "nclust = 12\n",
    "\n",
    "gp_eval_path = \"false\"\n",
    "gp_truth = \"false\"\n",
    "\n",
    "cond_keys = [\"noMiss_nmulti\",\"multiMissHS_nmulti\",\"rnaOnlyHS_nmulti\",\"atacOnlyHS_nmulti\",\n",
    "             \"multiMissEndo_nmulti\",\"rnaOnlyEndo_nmulti\",\"atacOnlyEndo_nmulti\"]\n",
    "\n",
    "iter_list = [3000]\n",
    "\n",
    "repeats = 5\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Unpaired + seurat 4\n",
    "\n",
    "# all R functions \n",
    "conda_envs = [\"seurat\",\"bindsc\",\"seurat\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3_single_noPred.R\",\n",
    "                  \"run_rbindsc_single_noPred.R\",\n",
    "                  \"run_seurat4_3_noPred.R\",\n",
    "                  \"run_rfigr_single.R\",\n",
    "                  \"run_rliger_single.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*5\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"seurat4/seurat4_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"seurat4\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=False, wait_time=1*30,\n",
    "                  batch=1,output_folder=\"results_single_mod\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "### Unpaired (multiome-split)\n",
    "\n",
    "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
    "# python script for running the method\n",
    "method_scripts =  [\"run_seurat3_noPred.R\",\n",
    "                   \"run_rbindsc_noPred.R\",\n",
    "                   \"run_rfigr_2.R\",\n",
    "                   \"run_rliger.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False]*4\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=False, wait_time=1*30,\n",
    "                  batch=1,output_folder=\"results_single_same_cell_number\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "### python functions\n",
    "conda_envs = [\"scmomat\",\"cobolt\",\"glue2\",\"multivi\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_scmomat_mouse_skin.py\",\n",
    "                  \"run_cobolt.py\",\n",
    "                  \"run_glue_mm10_single_noPred.py\",\n",
    "                  \"run_multivi_2_noPred.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]*4\n",
    "\n",
    "file_paths = [\"scmomat/scmomat_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"glue/glue_result.csv\",\n",
    "              \"multivi/multivi_result.csv\"]\n",
    "\n",
    "method_keys = [\"scmomat\",\"cobolt\",\"glue\",\"multivi\"]\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=True, wait_time=1*30,\n",
    "                  batch=4,output_folder=\"results_single_mod\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "\n",
    "\n",
    "conda_envs = [\"glue2\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_glue_mm10_noPred.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True]\n",
    "\n",
    "file_paths = [\"glue/glue_result.csv\"]\n",
    "\n",
    "method_keys = [\"glue\"]\n",
    "\n",
    "\n",
    "for i in range(len(cond_keys)):\n",
    "    cond_key = cond_keys[i]\n",
    "    eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "                  method_scripts,py_langs,file_paths,method_keys,\n",
    "                  ct_ref,nclust,dir_path,cond_key,\n",
    "                  iter_list,repeats,repeat_start=1,\n",
    "                  wait=True, wait_time=1*30,\n",
    "                  batch=2,output_folder=\"results_single_same_cell_number\",\n",
    "                  ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,rare_ct_path = rare_ct_path)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HPAP integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# working_dir \n",
    "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
    "# bash script for job submission\n",
    "job_submission_script = \"/home/myylee/scmint/methods_eval/submit_job_per_condition_n_eval2.sh\"\n",
    "# python script for metric evaluation \n",
    "eval_script = \"/home/myylee/scmint/methods_eval/run_metric_eval_batch2_hpap.py\"\n",
    "\n",
    "ct_ref = \"/project/mingyaolpc/myylee/scmint/methods_eval/dataset/hpap/hpap_all_bc_ct3.csv\"\n",
    "nclust = 10\n",
    "\n",
    "dir_path = \"/project/mingyaolpc/myylee/scmint/methods_eval/dataset/hpap/real_data/\"\n",
    "cond_key = \"all\"\n",
    "\n",
    "iter_list = [70000]\n",
    "repeats = 1\n",
    "\n",
    "gp_eval_path = \"false\"\n",
    "gp_truth = \"false\"\n",
    "mem_limit = 100\n",
    "\n",
    "# ===== functions ===== \n",
    "\n",
    "# == Python functions ==\n",
    "\n",
    "conda_envs = [\"multivi\",\"glue2\",\"cobolt\",\"scmomat\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_multivi_batch.py\",\n",
    "                 \"run_glue_hg38_batch.py\",\n",
    "                 \"run_cobolt.py\",\n",
    "                 \"run_scmomat_batch_hpap.py\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [True,True,True,True]\n",
    "\n",
    "file_paths = [\"multivi/multivi_result.csv\",\n",
    "              \"glue/glue_result.csv\",\n",
    "              \"cobolt/cobolt_result.csv\",\n",
    "              \"scmomat/scmomat_result.csv\"]\n",
    "\n",
    "method_keys = [\"multivi\",\"glue\",\"cobolt\",\"scmomat\"]\n",
    "\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=1,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,\n",
    "              mem_limit = mem_limit)\n",
    "\n",
    "# == R functions ==\n",
    "conda_envs = [\"seurat\",\"figr\",\"liger\",\"seurat\",\"bindsc\"]\n",
    "# python script for running the method\n",
    "method_scripts = [\"run_seurat3_batch_sequential.R\",\n",
    "                  \"run_rfigr_2_hpap_sequential.R\",\n",
    "                  \"run_rliger_batch_sequential.R\",\n",
    "                  \"run_seurat4_4_sequential.R\",\n",
    "                  \"run_rbindsc_batch_sequential.R\"]\n",
    "# if the script should be run in python environment \n",
    "py_langs = [False,False,False,False,False]\n",
    "\n",
    "file_paths = [\"seurat3/seurat3_result.csv\",\n",
    "              \"rfigr/rfigr_result.csv\",\n",
    "              \"rliger/rliger_result.csv\",\n",
    "              \"seurat4int/seurat4int_result.csv\",\n",
    "              \"rbindsc/rbindsc_result.csv\"]\n",
    "\n",
    "method_keys = [\"seurat3\",\"rfigr\",\"rliger\",\"seurat4int\",\"rbindsc\"]\n",
    "\n",
    "\n",
    "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
    "              method_scripts,py_langs,file_paths,method_keys,\n",
    "              ct_ref,nclust,dir_path,cond_key,\n",
    "              iter_list,repeats,repeat_start=1,\n",
    "              wait=True, wait_time=2*60,batch=1,output_folder=\"results\",\n",
    "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth,\n",
    "              mem_limit = mem_limit)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Copy of evaluate_pbmc_vary_depth.ipynb",
   "provenance": [
    {
     "file_id": "1awN0s01CNXA6Z4DJiqgfzkm6o6gHB_gO",
     "timestamp": 1647299165044
    }
   ]
  },
  "kernelspec": {
   "display_name": "scib2_test",
   "language": "python",
   "name": "scib2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "12px",
    "width": "256px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "687px",
    "left": "35px",
    "top": "111.141px",
    "width": "546.047px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
