{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/myylee/benchmark_sc_multiomic_integration/blob/main/evaluate_vary_situations_public.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rXhERnIMrgZx"
      },
      "outputs": [],
      "source": [
        "# Run with a python kernel"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "JFmJ7_9TrgZz"
      },
      "outputs": [],
      "source": [
        "%load_ext autoreload\n",
        "%autoreload 2\n",
        "\n",
        "# load libraries\n",
        "import h5py\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import scanpy as sc\n",
        "from anndata import AnnData\n",
        "import anndata as ad\n",
        "from copy import deepcopy\n",
        "from scipy.sparse import csr_matrix\n",
        "from scipy.sparse.linalg import svds\n",
        "import pickle\n",
        "import scipy.io as sio\n",
        "import os\n",
        "\n",
        "# Path to R [modify as needed]\n",
        "os.environ['R_HOME'] = '/home/myylee/anaconda3/envs/r_py/lib/R/'\n",
        "import utils_eval\n",
        "\n",
        "\n",
        "from data_simulation import pair_unpair_split_size,downsample_samples, data_simulation, data_simulation_batch\n",
        "from data_simulation import eval_test_all\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EYqoK-box7o5"
      },
      "source": [
        "# Sceanrio 1"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_6CBBlSKx7o6"
      },
      "source": [
        "## PBMC "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VAPZ5xm5x7o6"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "scrolled": true,
        "id": "VUFto-uSx7o6"
      },
      "outputs": [],
      "source": [
        "source_dir = \"dataset/multiome_pbmc_10k/\"\n",
        "\n",
        "adata_rna = ad.read_h5ad(source_dir+\"pbmc_10x_rna_public.h5ad\")\n",
        "adata_atac = ad.read_h5ad(source_dir+\"pbmc_10x_atac_public.h5ad\")\n",
        "\n",
        "in_dir = \"dataset/multiome_pbmc_10k/pbmc_vary_cell_test/\"\n",
        "\n",
        "iter_list = [1000,3000,8000]\n",
        "\n",
        "depth_multiome_list = [1]*len(iter_list)\n",
        "depth_scrna_list = [1]*len(iter_list)\n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "\n",
        "n_scrna_list = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = iter_list\n",
        "\n",
        "repeats = 5\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "fragment_path = \"{}/dataset/multiome_pbmc_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "\n",
        "cond_key = \"nmulti\"\n",
        "    \n",
        "def to_str(s): return(str(s))\n",
        "\n",
        "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
        "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
        "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
        "                repeats,fragment_path,cond_key,\n",
        "                to_str, downsample=False)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "C1BIugTNx7o6"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "syk-ZnSjx7o6"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_single.py\"\n",
        "\n",
        "ct_ref = \"dataset/multiome_pbmc_10k/pbmc_10x_bc_ct3.csv\"\n",
        "nclust = 7\n",
        "\n",
        "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
        "gp_truth = \"dataset/multiome_pbmc_10k/pbmc_10x_pmat_sig_links_50kb_unique.csv\"\n",
        "\n",
        "dir_path = \"dataset/multiome_pbmc_10k/pbmc_vary_cell_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [1000,3000,8000]\n",
        "repeats = 5\n",
        "\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired =====\n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3_single.R\",\n",
        "                  \"run_rbindsc_single.R\",\n",
        "                  \"run_rfigr_single.R\",\n",
        "                  \"run_rliger_single.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rliger/rliger_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"seurat4\",\"rfigr\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "R2si7ryvx7o7"
      },
      "source": [
        "### Add dashed line in peak-gene recovery percentage with the true paired RNA and ATAC profiles"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "1xvio2-xx7o8"
      },
      "outputs": [],
      "source": [
        "# copy over the true paired RNA-seq data\n",
        "import shutil\n",
        "dir_path = \"dataset/multiome_pbmc_10k/pbmc_vary_cell_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [1000,3000,8000]\n",
        "repeats = 5\n",
        "output_folder=\"results_single_same_cell_number\"\n",
        "\n",
        "for i in range(len(iter_list)):\n",
        "    for j in range(1,repeats+1):\n",
        "        in_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j))\n",
        "        out_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j),output_folder)\n",
        "\n",
        "        os.makedirs(os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\"), exist_ok=True)\n",
        "        # save ATAC to the right location\n",
        "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"), \n",
        "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"barcodes.tsv\"))\n",
        "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"peak.tsv\"), \n",
        "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"peak.tsv\"))\n",
        "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"ATAC_counts.mtx\"), \n",
        "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"counts.mtx\"))\n",
        "\n",
        "        # load in ATAC barcodes \n",
        "        atac_bc = pd.read_csv(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"),header=None)\n",
        "\n",
        "        # get RNA profile of the corresponding cells \n",
        "        in_dir = \"dataset/multiome_pbmc_10k/\"\n",
        "\n",
        "        adata_rna = ad.read_h5ad(in_dir+\"pbmc_10x_rna_public.h5ad\")\n",
        "\n",
        "        idx_sel = atac_bc[0].tolist()\n",
        "        # RNA\n",
        "        adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna[idx_sel,:].X.todense())),\n",
        "                               obs=adata_rna.obs.loc[idx_sel,:],\n",
        "                               var=adata_rna.var,dtype=np.float32)\n",
        "        adata_rna_sel.obs['rna.bc'] = adata_rna_sel.obs_names\n",
        "        # save RNA to the right location\n",
        "        utils_eval.write_adata(adata_rna_sel, os.path.join(out_dir_i,\"truth\",\"predicted\",\"RNA\"),\"RNA\",\"gene\",feature_name='feature',transpose=True)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "80AKRm9Bx7o8"
      },
      "outputs": [],
      "source": [
        "# run eval_missming_modality_prediction \n",
        "\n",
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"/home/myylee/scmint/methods_eval/submit_job_per_missing_mod_eval.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"/home/myylee/scmint/methods_eval/run_metric_eval_single.py\"\n",
        "\n",
        "ct_ref = \"dataset/multiome_pbmc_10k/pbmc_10x_bc_ct3.csv\"\n",
        "nclust = 7\n",
        "\n",
        "dir_path = \"dataset/multiome_pbmc_10k/pbmc_vary_cell_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [1000,3000,8000]\n",
        "repeats = 5\n",
        "\n",
        "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
        "gp_truth = \"dataset/multiome_pbmc_10k/pbmc_10x_pmat_sig_links_50kb_unique.csv\"\n",
        "method_keys = [\"truth\"]\n",
        "\n",
        "\n",
        "#----- All of these can be random strings, are not used in script, but need not to be empty strings ----\n",
        "conda_envs = [\"truth_na\"]\n",
        "# R script for running the method\n",
        "method_scripts = [\"run_truth.csv\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "file_paths = [\"truth/truth_result.csv\"]\n",
        "#------ END ----- \n",
        "\n",
        "\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ygc0gjBCx7o8"
      },
      "source": [
        "## BMMC"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "60SkXQ8Cx7o8"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cyKyZByVx7o8"
      },
      "outputs": [],
      "source": [
        "# using pmat dataset, generated from bmmc_process R script \n",
        "source_dir = \"dataset/bmmc/\"\n",
        "\n",
        "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
        "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "adata_atac.var['feature'] = adata_atac.var['features'] \n",
        "adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "\n",
        "# aggregate the selected donors \n",
        "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
        "\n",
        "# Paired - RNA\n",
        "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "# Paired - ATAC\n",
        "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "\n",
        "\n",
        "in_dir = \"dataset/bmmc/bmmc_vary_cell_test/\"\n",
        "iter_list = [1000,2000,4000]\n",
        "depth_multiome_list = [1]*len(iter_list)\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = iter_list\n",
        "repeats = 5\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
        "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "def to_str(s): return(str(s))\n",
        "\n",
        "\n",
        "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
        "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
        "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
        "                repeats,fragment_path,cond_key,\n",
        "                to_str, downsample=False)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "aw0NFijCx7o9"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hfT4inRVx7o9"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_single.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
        "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/bmmc_vary_cell_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [1000,2000,4000]\n",
        "repeats = 5\n",
        "\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired =====\n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3_single.R\",\n",
        "                  \"run_rbindsc_single.R\",\n",
        "                  \"run_rfigr_single.R\",\n",
        "                  \"run_rliger_single.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rliger/rliger_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"seurat4\",\"rfigr\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_mod\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mN6EE58Gx7o9"
      },
      "source": [
        "### Add dashed line in peak-gene recovery percentage with the true paired RNA and ATAC profiles"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "YgnnR8ntx7o9"
      },
      "outputs": [],
      "source": [
        "# copy over the true paired RNA-seq data\n",
        "import shutil\n",
        "dir_path = \"dataset/bmmc/bmmc_vary_cell_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [1000,2000,4000]\n",
        "repeats = 5\n",
        "output_folder=\"results_single_same_cell_number\"\n",
        "\n",
        "for i in range(len(iter_list)):\n",
        "    for j in range(1,repeats+1):\n",
        "        in_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j))\n",
        "        out_dir_i = os.path.join(dir_path,\"{}{}_{}\".format(cond_key,iter_list[i],j),output_folder)\n",
        "\n",
        "        os.makedirs(os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\"), exist_ok=True)\n",
        "        # save ATAC to the right location\n",
        "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"), \n",
        "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"barcodes.tsv\"))\n",
        "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"peak.tsv\"), \n",
        "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"peak.tsv\"))\n",
        "        shutil.copy2(os.path.join(in_dir_i,\"unpaired_ATAC\",\"ATAC_counts.mtx\"), \n",
        "                     os.path.join(out_dir_i,\"truth\",\"predicted\",\"ATAC\",\"counts.mtx\"))\n",
        "\n",
        "\n",
        "        # load in ATAC barcodes \n",
        "        atac_bc = pd.read_csv(os.path.join(in_dir_i,\"unpaired_ATAC\",\"barcodes.tsv\"),header=None)\n",
        "        idx_sel = atac_bc[0].tolist()\n",
        "\n",
        "        # get RNA profile of the corresponding cells \n",
        "        in_dir = \"dataset/bmmc/\"\n",
        "\n",
        "        adata_rna = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "        adata_rna.obs['rna.bc'] = adata_rna.obs_names\n",
        "        adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "        # RNA\n",
        "        adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna[idx_sel,:].X.todense())),\n",
        "                               obs=adata_rna.obs.loc[idx_sel,:],\n",
        "                               var=adata_rna.var,dtype=np.float32)\n",
        "        # save RNA to the right location\n",
        "        utils_eval.write_adata(adata_rna_sel, os.path.join(out_dir_i,\"truth\",\"predicted\",\"RNA\"),\"RNA\",\"gene\",feature_name='feature',transpose=True)\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "97Qhea9zx7o9"
      },
      "outputs": [],
      "source": [
        "# run eval_missming_modality_prediction \n",
        "\n",
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_single.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
        "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/bmmc_vary_cell_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [1000,2000,4000]\n",
        "repeats = 5\n",
        "\n",
        "method_keys = [\"truth\"]\n",
        "\n",
        "#----- All of these can be random strings, are not used in script, but need not to be empty strings ----\n",
        "conda_envs = [\"truth_na\"]\n",
        "# R script for running the method\n",
        "method_scripts = [\"run_truth.csv\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "file_paths = [\"truth/truth_result.csv\"]\n",
        "#------ END ----- \n",
        "\n",
        "\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results_single_same_cell_number\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TbzSfyipx7o-"
      },
      "source": [
        "# Sceanrio 2 "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RkhyL-i0x7o-"
      },
      "source": [
        "## PBMC 2000 cells "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KseFCnNGx7o-"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "DKsME0Qnx7o-"
      },
      "outputs": [],
      "source": [
        "source_dir = \"dataset/multiome_pbmc_10k/\"\n",
        "\n",
        "adata_rna = ad.read_h5ad(source_dir+\"pbmc_10x_rna_public.h5ad\")\n",
        "adata_atac = ad.read_h5ad(source_dir+\"pbmc_10x_atac_public.h5ad\")\n",
        "\n",
        "\n",
        "in_dir = \"dataset/multiome_pbmc_10k/nmulti2000_7ct_vdepth_test/\"\n",
        "iter_list = [0.25,0.5,0.75,1]\n",
        "depth_multiome_list = iter_list\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = [2000]*len(iter_list)\n",
        "repeats = 5\n",
        "\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "fragment_path = \"{}/dataset/multiome_pbmc_10k/pbmc_granulocyte_sorted_10k_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "\n",
        "cond_key = \"depthmulti\"\n",
        "\n",
        "def str_norm(s):\n",
        "    return(str(int(s*100)))\n",
        "\n",
        "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
        "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
        "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
        "                repeats,fragment_path,cond_key,\n",
        "                str_norm, downsample=True)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "akT12IULx7o-"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KcyShGfJx7o-"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_fair.py\"\n",
        "\n",
        "ct_ref = \"dataset/multiome_pbmc_10k/pbmc_10x_bc_ct3.csv\"\n",
        "nclust = 7\n",
        "\n",
        "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
        "gp_truth = \"dataset/multiome_pbmc_10k/pbmc_10x_pmat_sig_links_50kb_unique.csv\"\n",
        "\n",
        "dir_path = \"dataset/multiome_pbmc_10k/nmulti2000_7ct_vdepth_test/\"\n",
        "cond_key = \"depthmulti\"\n",
        "\n",
        "iter_list = [25,50,75,100]\n",
        "repeats = 5\n",
        "\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PWTCU607x7o-"
      },
      "source": [
        "## BMMC 2000 cells "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jDiQ4iYYx7o-"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BCvFBMVQx7o-"
      },
      "outputs": [],
      "source": [
        "# using pmat dataset, generated from bmmc_process R script \n",
        "in_dir = \"dataset/bmmc/\"\n",
        "\n",
        "adata_atac = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
        "adata_rna = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "adata_atac.var['feature'] = adata_atac.var['features'] \n",
        "adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "# aggregate the selected donors \n",
        "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
        "\n",
        "# Paired - RNA\n",
        "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "# Paired - ATAC\n",
        "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "\n",
        "# ensure that var_names are feature names and obs_names are cell barcodes\n",
        "adata_rna_sel.var_names = adata_rna.var['feature'].tolist()\n",
        "adata_atac_sel.var_names = adata_atac.var['feature'].tolist()\n",
        "\n",
        "in_dir = \"dataset/bmmc/nmulti2000_21ct_vdepth_test/\"\n",
        "iter_list = [0.25]#[0.25,0.5,0.75,1]\n",
        "depth_multiome_list = iter_list\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = [2000]*len(iter_list)\n",
        "repeats = 2 #5\n",
        "\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
        "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "cond_key = \"depthmulti\"\n",
        "\n",
        "def str_norm(s):\n",
        "    return(str(int(s*100)))\n",
        "\n",
        "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
        "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
        "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
        "                repeats,fragment_path,cond_key,\n",
        "                str_norm, downsample=True)\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "CuPu2oT5x7o-"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7etHYcx1x7o_"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_fair.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
        "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/nmulti2000_21ct_vdepth_test/\"\n",
        "cond_key = \"depthmulti\"\n",
        "\n",
        "iter_list = [25,50,75,100]\n",
        "repeats = 5\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Sk5idDRGx7o_"
      },
      "source": [
        "## BMMC 4000 cells "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "d2BsB4a0x7o_"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "pvZ97p-wx7o_"
      },
      "outputs": [],
      "source": [
        "# using pmat dataset, generated from bmmc_process R script \n",
        "in_dir = \"dataset/bmmc/\"\n",
        "\n",
        "adata_atac = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
        "adata_rna = ad.read_h5ad(in_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "adata_atac.var['feature'] = adata_atac.var['features'] \n",
        "adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "# aggregate the selected donors \n",
        "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
        "\n",
        "# Paired - RNA\n",
        "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "# Paired - ATAC\n",
        "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "\n",
        "# ensure that var_names are feature names and obs_names are cell barcodes\n",
        "adata_rna_sel.var_names = adata_rna.var['feature'].tolist()\n",
        "adata_atac_sel.var_names = adata_atac.var['feature'].tolist()\n",
        "\n",
        "in_dir = \"dataset/bmmc/nmulti4000_21ct_vdepth_test/\"\n",
        "iter_list = [0.25,0.5,0.75,1]\n",
        "depth_multiome_list = iter_list\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = [4000]*len(iter_list)\n",
        "repeats = 5\n",
        "\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
        "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "cond_key = \"depthmulti\"\n",
        "\n",
        "def str_norm(s):\n",
        "    return(str(int(s*100)))\n",
        "\n",
        "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
        "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
        "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
        "                repeats,fragment_path,cond_key,\n",
        "                str_norm, downsample=True)\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IcQ1X-__x7o_"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "gawrotjux7o_"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_fair.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
        "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/nmulti4000_21ct_vdepth_test/\"\n",
        "cond_key = \"depthmulti\"\n",
        "\n",
        "iter_list = [25,50,75,100]\n",
        "repeats = 5\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OElJraF9x7o_"
      },
      "source": [
        "## Increasing number of cells; 100% depth"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tyZIbpE0x7o_"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BcVwSdVCx7o_"
      },
      "outputs": [],
      "source": [
        "source_dir = \"dataset/bmmc/\"\n",
        "\n",
        "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
        "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "adata_atac.var['feature'] = adata_atac.var['features'] \n",
        "adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "\n",
        "# aggregate the selected donors \n",
        "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
        "\n",
        "# Paired - RNA\n",
        "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "# Paired - ATAC\n",
        "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "\n",
        "# ensure that var_names are feature names and obs_names are cell barcodes\n",
        "adata_rna_sel.var_names = adata_rna.var['feature'].tolist()\n",
        "adata_atac_sel.var_names = adata_atac.var['feature'].tolist()\n",
        "\n",
        "\n",
        "in_dir = \"dataset/bmmc/bmmc_vcells_intervals/\"\n",
        "iter_list = [int(x) for x in np.linspace(start=1000, stop=1000+400*10, num=11)] # ncell increase interval, each interval: increase by 400 cells. 400 cells of 10k atac fragment and 2500 nCounts_RNA is roughly equal to 10% increment in ATAC and RNA sequencing depth, assuming 4000 cells\n",
        "depth_multiome_list = [1]*len(iter_list)\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = iter_list\n",
        "repeats = 5\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
        "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "cond_key = \"nmulti\"\n",
        "    \n",
        "def to_str(s): return(str(s))\n",
        "\n",
        "\n",
        "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
        "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
        "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
        "                repeats,fragment_path,cond_key,\n",
        "                to_str, downsample=False)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zDxspTZEx7pA"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "m9xO8Yrnx7pA"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_fair.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
        "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/bmmc_vcells_intervals/\"\n",
        "cond_key = \"depthmulti\"\n",
        "\n",
        "iter_list = [25,50,75,100]\n",
        "repeats = 5\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YWuUTYlNx7pA"
      },
      "source": [
        "## Increasing depth; 4000 cells"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ziN1ZMsZx7pA"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Ahjwx--Sx7pA"
      },
      "outputs": [],
      "source": [
        "source_dir = \"dataset/bmmc/\"\n",
        "\n",
        "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
        "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "adata_atac.var['feature'] = adata_atac.var['features'] \n",
        "adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "# aggregate the selected donors \n",
        "idx_sel = list(adata_atac.obs['batch'].isin([\"s1d2\"]))\n",
        "\n",
        "# Paired - RNA\n",
        "adata_rna_sel = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_sel,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_sel,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "# Paired - ATAC\n",
        "adata_atac_sel = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_sel,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_sel,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "\n",
        "# ensure that var_names are feature names and obs_names are cell barcodes\n",
        "adata_rna_sel.var_names = adata_rna.var['feature'].tolist()\n",
        "adata_atac_sel.var_names = adata_atac.var['feature'].tolist()\n",
        "\n",
        "\n",
        "# Saving everything in /project/mingyaolpc/ folder to avoid running out of storage\n",
        "in_dir = \"dataset/bmmc/bmmc_vdepth_intervals/\"\n",
        "iter_list = np.linspace(start=10, stop=100, num=10)/100\n",
        "depth_multiome_list = iter_list\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = [4000]*len(iter_list)\n",
        "repeats = 5\n",
        "\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
        "fragment_path = \"{}/dataset/bmmc/s1d2_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "cond_key = \"depthmulti\"\n",
        "\n",
        "def str_norm(s):\n",
        "    return(str(int(s*100)))\n",
        "\n",
        "data_simulation(in_dir,adata_rna_sel,adata_atac_sel,iter_list,\n",
        "                depth_multiome_list,depth_scrna_list,depth_snatac_list,\n",
        "                n_multiome_list,n_scrna_list,n_snatac_list,\n",
        "                repeats,fragment_path,cond_key,\n",
        "                str_norm, downsample=True)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "x7U2H1-rx7pA"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "DKh7tauxx7pA"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_fair.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"eval_missing_modality_prediction_single.R\"\n",
        "gp_truth = \"dataset/bmmc/bmmc_s1d2_pmat_all_ct_sig_links_50kb_unique.csv\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/bmmc_vdepth_intervals/\"\n",
        "cond_key = \"depthmulti\"\n",
        "\n",
        "iter_list = [25,50,75,100]\n",
        "repeats = 5\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_2.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GK_IiOI2x7pB"
      },
      "source": [
        "# Sceanrio 3"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IGEyNteMx7pB"
      },
      "source": [
        "## Technical batch effect "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VGKbmZpHx7pB"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "mLeRiSOVx7pB"
      },
      "outputs": [],
      "source": [
        "# using pmat dataset, generated from bmmc_process R script \n",
        "source_dir = \"dataset/bmmc/\"\n",
        "\n",
        "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
        "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "adata_atac.var['feature'] = adata_atac.var['features'] \n",
        "adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "# Paired\n",
        "idx_s1d1 = list(adata_atac.obs['batch'] == 's1d1')\n",
        "\n",
        "# Paired - RNA\n",
        "adata_rna_s1d1 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s1d1,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_s1d1,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "# Paired - ATAC\n",
        "adata_atac_s1d1 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s1d1,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_s1d1,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "# snRNA\n",
        "idx_s2d1 = list(adata_atac.obs['batch'] == 's2d1')\n",
        "adata_rna_s2d1 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s2d1,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_s2d1,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "\n",
        "# snATAC\n",
        "adata_atac_s2d1 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s2d1,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_s2d1,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "\n",
        "in_dir = \"dataset/bmmc/bmmc_technical_batch_test/\"\n",
        "iter_list = [1000,3000,5000]\n",
        "depth_multiome_list = [1]*len(iter_list)\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = iter_list\n",
        "repeats = 5\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
        "fragment_path = \"{}/dataset/bmmc/bmmc_site1_or_donor1_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "cond_key = \"nmulti\"\n",
        "    \n",
        "def to_str(s): return(str(s))\n",
        "\n",
        "# assume there are ['rna.bc','batch'] in each adata.obs\n",
        "# assume there is ['feature'] in each adata.var\n",
        "data_simulation_batch(in_dir,adata_rna_s1d1,adata_atac_s1d1,adata_rna_s2d1,adata_atac_s2d1,\n",
        "                      iter_list,depth_multiome_list,depth_scrna_list,\n",
        "                      depth_snatac_list,n_multiome_list,\n",
        "                      n_scrna_list,n_snatac_list,\n",
        "                      repeats,fragment_path,cond_key,\n",
        "                      to_str, downsample=False,same_unpair_origin=True)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "S6Xxwspqx7pB"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "zpS18dL1x7pB"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_batch.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"false\"\n",
        "gp_truth = \"false\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/bmmc_technical_batch_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [1000,3000,5000]\n",
        "repeats = 5\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_batch.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "7x7Z3jORx7pB"
      },
      "source": [
        "## Biological batch effect "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "neBIeEYAx7pB"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dY1MIHF6x7pB"
      },
      "outputs": [],
      "source": [
        "# using pmat dataset, generated from bmmc_process R script \n",
        "source_dir = \"dataset/bmmc/\"\n",
        "\n",
        "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
        "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "adata_atac.var['feature'] = adata_atac.var['features'] \n",
        "adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "# Paired\n",
        "idx_s1d1 = list(adata_atac.obs['batch'] == 's1d1')\n",
        "\n",
        "# Paired - RNA\n",
        "adata_rna_s1d1 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s1d1,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_s1d1,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "# Paired - ATAC\n",
        "adata_atac_s1d1 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s1d1,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_s1d1,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "# snRNA\n",
        "idx_s1d2 = list(adata_atac.obs['batch'] == 's1d2')\n",
        "adata_rna_s1d2 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s1d2,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_s1d2,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "\n",
        "# snATAC\n",
        "adata_atac_s1d2 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s1d2,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_s1d2,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "\n",
        "in_dir = \"dataset/bmmc/bmmc_biological_batch_test/\"\n",
        "iter_list = [1000,3000,5000]\n",
        "depth_multiome_list = [1]*len(iter_list)\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = iter_list\n",
        "repeats = 5\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
        "fragment_path = \"{}/dataset/bmmc/bmmc_site1_or_donor1_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "cond_key = \"nmulti\"\n",
        "    \n",
        "def to_str(s): return(str(s))\n",
        "\n",
        "# assume there are ['rna.bc','batch'] in each adata.obs\n",
        "# assume there is ['feature'] in each adata.var\n",
        "data_simulation_batch(in_dir,adata_rna_s1d1,adata_atac_s1d1,adata_rna_s2d1,adata_atac_s2d1,\n",
        "                      iter_list,depth_multiome_list,depth_scrna_list,\n",
        "                      depth_snatac_list,n_multiome_list,\n",
        "                      n_scrna_list,n_snatac_list,\n",
        "                      repeats,fragment_path,cond_key,\n",
        "                      to_str, downsample=False,same_unpair_origin=True)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wEL5iu8Kx7pC"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "hYV9GAfHx7pC"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_batch.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"false\"\n",
        "gp_truth = \"false\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/bmmc_biological_batch_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [1000,3000,5000]\n",
        "repeats = 5\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_batch.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "L8nqntILx7pC"
      },
      "source": [
        "## Complex test #1"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rynZFn9Yx7pC"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "VEzkspnNx7pC"
      },
      "outputs": [],
      "source": [
        "# using pmat dataset, generated from bmmc_process R script \n",
        "source_dir = \"dataset/bmmc/\"\n",
        "\n",
        "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
        "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "adata_atac.var['feature'] = adata_atac.var['features'] \n",
        "adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "# Paired\n",
        "idx_paired = list(adata_atac.obs['batch'].isin(['s1d1','s1d3']))\n",
        "\n",
        "# Paired - RNA\n",
        "adata_rna_paired = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_paired,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_paired,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "# Paired - ATAC\n",
        "adata_atac_paired = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_paired,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_paired,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "# snRNA\n",
        "idx_s1d2 = list(adata_atac.obs['batch'] == 's1d2')\n",
        "adata_rna_s1d2 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s1d2,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_s1d2,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "\n",
        "# snATAC\n",
        "adata_atac_s1d2 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s1d2,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_s1d2,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "\n",
        "in_dir = \"dataset/bmmc/bmmc_complex1_test/\"\n",
        "iter_list = [1000,3000,5000]\n",
        "depth_multiome_list = [1]*len(iter_list)\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [1000]*len(iter_list)\n",
        "n_snatac_list = [1000]*len(iter_list)\n",
        "n_multiome_list = iter_list\n",
        "repeats = 5\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
        "fragment_path = \"{}/dataset/bmmc/bmmc_site1_or_donor1_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "cond_key = \"nmulti\"\n",
        "    \n",
        "def to_str(s): return(str(s))\n",
        "\n",
        "# assume there are ['rna.bc','batch'] in each adata.obs\n",
        "# assume there is ['feature'] in each adata.var\n",
        "data_simulation_batch(in_dir,adata_rna_s1d1,adata_atac_s1d1,adata_rna_s2d1,adata_atac_s2d1,\n",
        "                      iter_list,depth_multiome_list,depth_scrna_list,\n",
        "                      depth_snatac_list,n_multiome_list,\n",
        "                      n_scrna_list,n_snatac_list,\n",
        "                      repeats,fragment_path,cond_key,\n",
        "                      to_str, downsample=False,same_unpair_origin=True)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yVJegkFIx7pC"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Jl4nCFItx7pC"
      },
      "outputs": [],
      "source": [
        "# working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_batch.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"false\"\n",
        "gp_truth = \"false\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/bmmc_complex1_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [1000,3000,5000]\n",
        "repeats = 5\n",
        "\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_batch.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\",\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\",\"run_seurat4_4.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*2\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\",\"seurat4int/seurat4int_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4int\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "bCCjiAU8x7pC"
      },
      "source": [
        "## Complex test #2"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eDA6b3F9x7pC"
      },
      "source": [
        "### Data simulation"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BFVDIHeXx7pC"
      },
      "outputs": [],
      "source": [
        "# using pmat dataset, generated from bmmc_process R script \n",
        "source_dir = \"dataset/bmmc/\"\n",
        "\n",
        "adata_atac = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_ATAC_pmat.h5ad\")\n",
        "adata_rna = ad.read_h5ad(source_dir+\"multiome_bmmc_site1_or_donor1_RNA.h5ad\")\n",
        "\n",
        "adata_atac.var['feature'] = adata_atac.var['features'] \n",
        "adata_rna.var['feature'] = adata_rna.var['features'] \n",
        "\n",
        "# Paired\n",
        "idx_paired = list(adata_atac.obs['batch'].isin(['s4d1','s1d3']))\n",
        "\n",
        "# Paired - RNA\n",
        "adata_rna_paired = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_paired,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_paired,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "# Paired - ATAC\n",
        "adata_atac_paired = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_paired,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_paired,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "# snRNA\n",
        "idx_s2d1 = list(adata_atac.obs['batch'] == 's2d1')\n",
        "adata_rna_s2d1 = AnnData(csr_matrix(deepcopy(adata_rna.X[idx_s2d1,:].todense())),\n",
        "                       obs=adata_rna.obs.iloc[idx_s2d1,:],\n",
        "                       var=adata_rna.var,dtype=np.float32)\n",
        "\n",
        "# snATAC\n",
        "adata_atac_s2d1 = AnnData(csr_matrix(deepcopy(adata_atac.X[idx_s2d1,:].todense())),\n",
        "                       obs=adata_atac.obs.iloc[idx_s2d1,:],\n",
        "                       var=adata_atac.var,dtype=np.float32)\n",
        "\n",
        "in_dir = \"dataset/bmmc/bmmc_complex2_test/\"\n",
        "iter_list = [10000]\n",
        "depth_multiome_list = [1]*len(iter_list)\n",
        "depth_scrna_list = [1]*len(iter_list) \n",
        "depth_snatac_list = [1]*len(iter_list)\n",
        "n_scrna_list  = [2000]*len(iter_list)\n",
        "n_snatac_list = [2000]*len(iter_list)\n",
        "n_multiome_list = iter_list\n",
        "repeats = 5\n",
        "\n",
        "\n",
        "# if do not create fragment file symlink, use fragment_path = None \n",
        "# fragment file was created using correct_s1d2_fragments_sel.ipynb, where barcode was made the same as stored in adata  \n",
        "fragment_path = \"{}/dataset/bmmc/bmmc_site1_or_donor1_atac_fragments.tsv.gz\".format(os.getcwd())\n",
        "cond_key = \"nmulti\"\n",
        "    \n",
        "def to_str(s): return(str(s))\n",
        "\n",
        "# assume there are ['rna.bc','batch'] in each adata.obs\n",
        "# assume there is ['feature'] in each adata.var\n",
        "data_simulation_batch(in_dir,adata_rna_s1d1,adata_atac_s1d1,adata_rna_s2d1,adata_atac_s2d1,\n",
        "                      iter_list,depth_multiome_list,depth_scrna_list,\n",
        "                      depth_snatac_list,n_multiome_list,\n",
        "                      n_scrna_list,n_snatac_list,\n",
        "                      repeats,fragment_path,cond_key,\n",
        "                      to_str, downsample=False,same_unpair_origin=True)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "EBb2DOODx7pD"
      },
      "source": [
        "### Running method + evaluations"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cUg8U__rx7pD"
      },
      "outputs": [],
      "source": [
        "### working_dir \n",
        "folder_dir=\"/home/myylee/scmint/methods_eval/\"\n",
        "# bash script for job submission\n",
        "job_submission_script = \"submit_job_per_condition_n_eval2.sh\"\n",
        "# python script for metric evaluation \n",
        "eval_script = \"run_metric_eval_batch.py\"\n",
        "\n",
        "ct_ref = \"dataset/bmmc/bmmc_all_bc_ct3.csv\"\n",
        "nclust = 21\n",
        "\n",
        "gp_eval_path = \"false\"\n",
        "gp_truth = \"false\"\n",
        "\n",
        "dir_path = \"dataset/bmmc/bmmc_complex2_test/\"\n",
        "cond_key = \"nmulti\"\n",
        "\n",
        "iter_list = [10000]\n",
        "repeats = 5\n",
        "\n",
        "# ===== multiome-guided (python functions) =====\n",
        "\n",
        "conda_envs = [\"multivi\",\"cobolt\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_multivi_batch.py\",\"run_cobolt.py\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [True]*2 \n",
        "\n",
        "file_paths = [\"multivi/multivi_result.csv\",\n",
        "              \"cobolt/cobolt_result.csv\"]\n",
        "\n",
        "method_keys = [\"multivi\",\"cobolt\"]\n",
        "\n",
        "\n",
        "# submit 2 jobs at a time, wait, submit new jobs until the 2 were done\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=True, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== multiome-guided (R functions) =====\n",
        "\n",
        "conda_envs = [\"seurat\",\"seurat\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat4_3.R\",\"run_seurat4_4.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*2\n",
        "\n",
        "file_paths = [\"seurat4/seurat4_result.csv\",\"seurat4int/seurat4int_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat4int\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n",
        "\n",
        "# ===== unpaired (multiome-split) ===== \n",
        "conda_envs = [\"seurat\",\"bindsc\",\"figr\",\"liger\"]\n",
        "# python script for running the method\n",
        "method_scripts = [\"run_seurat3.R\",\"run_rbindsc.R\",\"run_rfigr_2.R\",\"run_rliger.R\"]\n",
        "# if the script should be run in python environment \n",
        "py_langs = [False]*4\n",
        "\n",
        "file_paths = [\"seurat3/seurat3_result.csv\",\n",
        "              \"rbindsc/rbindsc_result.csv\",\n",
        "              \"rfigr/rfigr_result.csv\",\n",
        "             \"rliger/rliger_result.csv\"]\n",
        "\n",
        "method_keys = [\"seurat3\",\"rbindsc\",\"rfigr\",\"rliger\"]\n",
        "\n",
        "# submit all jobs at once\n",
        "eval_test_all(folder_dir,job_submission_script,eval_script,conda_envs,\n",
        "              method_scripts,py_langs,file_paths,method_keys,\n",
        "              ct_ref,nclust,dir_path,cond_key,\n",
        "              iter_list,repeats,repeat_start=1,\n",
        "              wait=False, wait_time=2*60,batch=2,output_folder=\"results\",\n",
        "              ncore=8,gp_script=gp_eval_path,gp_truth=gp_truth)\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "scib2",
      "language": "python",
      "name": "scib2"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.7.12"
    },
    "toc": {
      "base_numbering": 1,
      "nav_menu": {
        "height": "12px",
        "width": "256px"
      },
      "number_sections": true,
      "sideBar": true,
      "skip_h1_title": false,
      "title_cell": "Table of Contents",
      "title_sidebar": "Contents",
      "toc_cell": false,
      "toc_position": {
        "height": "687px",
        "left": "35px",
        "top": "111.141px",
        "width": "546.047px"
      },
      "toc_section_display": true,
      "toc_window_display": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}